# Иванов Роман
# 98 Вариант
# Группа №9372
# Задача 1

Необходимо:
1. Найти новый рацион, удовлетворяющий только ограничению по питательным веществам.

2. Найти новый рацион, удовлетворяющий ограничениям по питательным веществам и выбросу парниковых газов.

3. Провести анализ чувствительности рациона к стоимости и порогу на выброс парниковых газов.


Пользователь является мужчиной 25 лет и занят легким физическим трудом.
Учесть также, что пользователь не планирует тратить более 9000 руб. в месяц на питание, а
средний выброс парниковых газов в день от производства продуктов, входящих в рацион, не
должен превышать 1,8 кг CO2-эквивалента.

Необходимо поставить задачу минимальной трансформации текущего рациона, при которой полученный рацион будет удовлетворять заданным требованиям Оценка трансформации («штраф») должна отталкиваться от абсолютного изменения количества продуктов, входящих в рацион, при этом должна также учитываться популярность продуктов. Желательным направлением изменения является увеличение потребления популярных продуктов и уменьшение — непопулярных в данной местности. Таким образом, «штраф» за увеличение количества некоторого продукта в рационе снижается с ростом популярности этого продукта. И наоборот — «штраф» за уменьшение количества продукта в рационе повышается с ростом его популярности.

![title](1.png)

Содержание питательныхх веществ в каждом продукте(https://ndb.nal.usda.gov/) и суточная норма(http://spravpit.liferus.ru/1_fiziol_normy.aspx)

![title](6.png)

## Формализация задачи и составление целевой функции

По условию задачи необходимо составить трансформацию исходного рациона. 
Таким образом, чтобы штраф» за увеличение количества некоторого продукта в рационе снижался с ростом популярности этого продукта. И наоборот — «штраф» за уменьшение количества продукта в рационе повышался с ростом его популярности.

Введем для каждого продукта две переменные: $x_i$$^+$ и $x_i$$^-$. Отвечающие за увеличение i-ого продукта в рационе и уменьшение соответственно. Кроме этого, необходимо учитывать популярность продукта в данном регионе. 

Тогда для каждой перменной введем коэффициенты: $(2-популярность)$ и $\frac{1}{2-популярность}$
Константа 2 введена для того, чтобы избавиться от нулевого штрафа.


Таким образом, наша целевая функция примет вид: 
$f(x)$ $=$ $(2-популярность)$*$x_1$$^+$+$\frac{1}{2-популярность}$*$x_1$$^-$+$...$+$(2-популярность)$*$x_i$$^+$+$\frac{1}{2-популярность}$*$x_i$$^-$+$...$+$(2-популярность)$*$x_1$$_3$$^+$$ $+$\frac{1}{2-популярность}$*$x_1$$_3$$^-$



# 1 Задание. Найти новый рацион, удовлетворяющий только ограничению по питательным веществам.

Для удобства напишем функцию, которая будет подсчитывать все коффициенты целевой функции.


In [3]:
popul=[1, 1, 0.95, 0.8, 0.8, 0.75, 0.78, 0.8, 0.75, 0.85, 0.65, 0.8, 0.9]

def calculCoef(array):
    result = []
    for i in range(len(array)):
        result.append((2-array[i])*1.0)
        result.append(1/(2-array[i]))
    return result

Так как нам необходимо трансформировать именно текущий рацион элементы функций ограничений будет иметь следующий вид.

($x_i$$^+$ - $x_1$$^-$)*содержание
т.е. например для картофеля:
($x_i$$^+$ - $x_1$$^-$)*$1.06$

Далле добавляются ограничение на то, чтобы трансформация не оказалась отрицательной

После расскрытия скобок у нас получится свободный член, подсчитаем его так же с помощью функции

In [4]:
#далле все функции прописываются с учетом знака <=

potreb = [112, 90, 120, 100, 53, 0, 143, 20, 154, 378, 15, 0, 317] #потредление в день до трансформации

stoim = [0.029, 0.0477, 0.747, 0.081, 0.3, 0.15, 0.3, 1.06, 0.56, 0.072, 1.25, 0.61, 0.53] # стоимость за грамм пробукта

nutrious = [[1.06, 4.34, 0.66, 0.17,2.32, 4, 0.71, 0, 0.37, 0.47, 3, 3.4, 0.3], #содержание питательных веществ в каждом продукте
            [0.2, 3.35, 5.13, 0.26, 2.36, 3, 1.2, 0, 3.89, 3.16, 4, 1.35, 1.74],
            [2.99, 0.75, 1.21, 0.41, 0.68, 3.91, 0.74, 0, 0.76, 0.5, 3.6, 6.47, 4.35],
            [14.4, 38, 0.64, 0.91, 29.4, 49, 8.51, 0, 0.82, 1.94, 62.2, 80.4, 53],
            [0, 0, 0.011, 0, 0.0019, 0, 0, 0, 0.015, 0.0016, 0, 0.028, 0.0264],
            [130, 4, 0, 46, 1, 0, 32, 0, 2.43, 160, 8, 37, 0],
            [0, 0, 1.43, 0.03, 0.07, 0, 2.14, 7.58, 0, 0.56, 0, 0.69, 0],
            [0.5, 3.3, 10.3, 1.8, 37, 13.1, 3.8, 0, 2.4, 1.8, 61.3, 11.4, 0],
            [0, 0, 0.022, 0, 0, 0, 0, 0.0004, 0.0005, 0.04, 0, 0.131, 0.002]]

nutriousNorm = [18, 21, 21, 200, 0.03, 750, 10, 150, 0.15] #норма потребления питательных веществ

def calculBForOgran(potreb, nutrious, stoim):
    
    #вычисление b для каждого ограничения по питательным веществам
    result = []
    for i in range(len(nutrious)):
        temp = 0
        for j in range(len(nutrious[i])):
            temp+=nutrious[i][j]*potreb[j]
        
        result.append(-nutriousNorm[i] + temp)
        
    #b для ограничения для того, чтобы потребление нашего клиента не уходили в минус
    #112 + x_i^+ - x_i^- >=0
    #-x_i^+ + x_i^- <= 112
    for i in range(len(potreb)):
        result.append(potreb[i])
        
    #ограничение на стоимость
    temp=0
    for i in range(len(stoim)):
        temp+=potreb[i]*stoim[i]
    result.append(333-temp)
    return result

def calculOgran(nutrious, stoim):
    #коеффициенты для ограничения по питательным веществам
    result = [[],[],[],[],[],[],[],[],[]]
    for i in range(len(nutrious)):
        for j in range(len(nutrious[i])):
            result[i].append(-nutrious[i][j])
            result[i].append(nutrious[i][j])
            
            #далее прописываются ограничения для того, чтобы потребление нашего клиента не уходили в минус
            #x_i^ - x_i^- >=0
            #-x_i^+ + x_i^- <= 112
    result.append([-1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,-1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,-1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,-1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,-1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,-1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,0,0,-1,1,0,0,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,1,0,0,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,1,0,0,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,1,0,0,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,1,0,0,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,1,0,0])
    result.append([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,1])
    
    #ограничение на стоимость
    result.append([])
    for i in range(len(stoim)):
        result[len(result)-1].append(stoim[i])
        result[len(result)-1].append(-stoim[i])
    return result

Теперь соберем все необходимые условия вместе

In [5]:
from scipy.optimize import linprog

In [6]:
c = calculCoef(popul)
A_ub = calculOgran(nutrious, stoim)
b_ub = calculBForOgran(potreb, nutrious, stoim)

In [7]:
print(linprog(c, A_ub, b_ub, method='simplex')['x'])
resultDiet = linprog(c, A_ub, b_ub, method='simplex')['x']

[  0.           0.           0.           0.           0.
 120.           0.           0.           0.           0.
   0.           0.           0.           0.           0.
  20.           0.          40.90535714   0.           0.
   0.          15.           0.           0.           0.
   0.        ]


In [8]:
count = 0
newPotreb = []
for i in range(len(potreb)):
    newPotreb.append(potreb[i]+resultDiet[count]-resultDiet[count+1])
    count+=2

In [9]:
print("Новый рацион с ограничениями по питательным веществам:")
label = ['Картофель', 'Хлеб','Яйца','Фрукты','Печенье','Злаки','Овощи','Масло','Сыр','Молоко','Орехи','Рыба','Мясо']
for i in range(len(label)):
    print(label[i] + ": " , newPotreb[i])

Новый рацион с ограничениями по питательным веществам:
Картофель:  112.0
Хлеб:  90.0
Яйца:  5.968558980384842e-13
Фрукты:  100.0
Печенье:  53.0
Злаки:  0.0
Овощи:  143.0
Масло:  0.0
Сыр:  113.09464285713449
Молоко:  378.0
Орехи:  3.552713678800501e-15
Рыба:  0.0
Мясо:  317.0


In [10]:
print("Старый рацион без ограничений:")
label = ['Картофель', 'Хлеб','Яйца','Фрукты','Печенье','Злаки','Овощи','Масло','Сыр','Молоко','Орехи','Рыба','Мясо']
for i in range(len(label)):
    print(label[i] + ": " , potreb[i])

Старый рацион без ограничений:
Картофель:  112
Хлеб:  90
Яйца:  120
Фрукты:  100
Печенье:  53
Злаки:  0
Овощи:  143
Масло:  20
Сыр:  154
Молоко:  378
Орехи:  15
Рыба:  0
Мясо:  317


In [11]:
sum = 0
for i in range(len(potreb)):
    sum+=potreb[i]*stoim[i]
print("Сумма в день, которую тратили ранее:", sum)

Сумма в день, которую тратили ранее: 485.49700000000007


In [12]:
sum = 0
for i in range(len(potreb)):
    sum+=newPotreb[i]*stoim[i]
print("Cумма в день, которую тратить будут сейчас:", sum)

Cумма в день, которую тратить будут сейчас: 332.9999999999958


Некоторые перменные получили очень маленькое значение, их можем считать за 0. И делаем вывод такой, что потребитель должен избавиться в своем рационе полостью от **яиц**, **масла**, **орехов**
Кроме этого он должен уменьшить количество потребляемого **сыра** чуть более, чем на **40гр**

# 2. Найти новый рацион, удовлетворяющий ограничениям по питательным веществам и выбросу парниковых газов.

In [13]:
# Проделываем ту же инициализацию
c = calculCoef(popul)
A_ub = calculOgran(nutrious, stoim)
b_ub = calculBForOgran(potreb, nutrious, stoim)

In [14]:
#добавляем ограничения по выбросам в массив b_ub
vibros = [0.0015, 0.0009, 0.0027, 0.0007, 0.0014, 0.0012, 0.0016, 0.0022, 0.0119, 0.0013, 0.0021, 0.0077, 0.0065]

temp = 0
for i in range(len(vibros)):
    temp+=potreb[i]*vibros[i]
b_ub.append(1.8 - temp)

A_ub.append([])
#добавляем ограничения по выбросам в массив A_ub
for i in range(len(vibros)):
    A_ub[len(A_ub)-1].append(vibros[i])
    A_ub[len(A_ub)-1].append(-vibros[i])

In [15]:
print(linprog(c, A_ub, b_ub, method='simplex')['x'])
resultDiet2 = linprog(c, A_ub, b_ub, method='simplex')['x']

[  0.           0.           0.           0.           0.
   0.           0.           0.           0.           0.
   0.           0.           0.           0.           0.
   0.           0.         154.           0.           0.
   0.           0.           0.           0.           0.
 272.83076923]


## Сравнение прошлых рационов и нового с ограничением на пит.вещества и выбросы

In [16]:
count = 0
newPotreb2 = []
for i in range(len(potreb)):
    newPotreb2.append(potreb[i]+resultDiet2[count]-resultDiet2[count+1])
    count+=2
print("Новый рацион с ограничениями по питательным веществам и выбросам:")
label = ['Картофель', 'Хлеб','Яйца','Фрукты','Печенье','Злаки','Овощи','Масло','Сыр','Молоко','Орехи','Рыба','Мясо']
for i in range(len(label)):
    print(label[i] + ": " , newPotreb2[i])
sum = 0
for i in range(len(potreb)):
    sum+=newPotreb2[i]*stoim[i]
print("Cумма в день, которую тратить будут с ограничениями на питательные вещества и выбросы:", sum)

Новый рацион с ограничениями по питательным веществам и выбросам:
Картофель:  112.0
Хлеб:  90.0
Яйца:  120.0
Фрукты:  100.0
Печенье:  53.0
Злаки:  0.0
Овощи:  143.0
Масло:  20.0
Сыр:  2.842170943040401e-14
Молоко:  378.0
Орехи:  15.0
Рыба:  0.0
Мясо:  44.16923076923524
Cумма в день, которую тратить будут с ограничениями на питательные вещества и выбросы: 254.65669230769473


In [17]:
print("Новый рацион с ограничениями по питательным веществам:")
label = ['Картофель', 'Хлеб','Яйца','Фрукты','Печенье','Злаки','Овощи','Масло','Сыр','Молоко','Орехи','Рыба','Мясо']
for i in range(len(label)):
    print(label[i] + ": " , newPotreb[i])
sum = 0
for i in range(len(potreb)):
    sum+=newPotreb[i]*stoim[i]
print("Cумма в день, которую тратить будут с ограничениями на питательные вещества:", sum)

Новый рацион с ограничениями по питательным веществам:
Картофель:  112.0
Хлеб:  90.0
Яйца:  5.968558980384842e-13
Фрукты:  100.0
Печенье:  53.0
Злаки:  0.0
Овощи:  143.0
Масло:  0.0
Сыр:  113.09464285713449
Молоко:  378.0
Орехи:  3.552713678800501e-15
Рыба:  0.0
Мясо:  317.0
Cумма в день, которую тратить будут с ограничениями на питательные вещества: 332.9999999999958


In [18]:
print("Старый рацион без ограничений:")
label = ['Картофель', 'Хлеб','Яйца','Фрукты','Печенье','Злаки','Овощи','Масло','Сыр','Молоко','Орехи','Рыба','Мясо']
for i in range(len(label)):
    print(label[i] + ": " , potreb[i])
sum = 0
for i in range(len(potreb)):
    sum+=potreb[i]*stoim[i]
print("Сумма в день, которую тратили при рационе без ограничений:", sum)

Старый рацион без ограничений:
Картофель:  112
Хлеб:  90
Яйца:  120
Фрукты:  100
Печенье:  53
Злаки:  0
Овощи:  143
Масло:  20
Сыр:  154
Молоко:  378
Орехи:  15
Рыба:  0
Мясо:  317
Сумма в день, которую тратили при рационе без ограничений: 485.49700000000007


Как мы можем заметить, при добавлении новых ограничений на выбросы в атмосферу, так же произошли некоторые изменения

|Продукт       |Без ограничений|Ограничения на пит.вещества|Ограничения на пит.вещества и выбросы| 
|--------------|---------------|---------------------------|-------------------------------------|
| **Картофель**|  112          | 112                       | 112                                 |
| **Хлеб**     |  90           | 90                        | 90                                  |
| **Яйца**     |  120          | 0                         | 120                                 |
| **Фрукты**   |  100          | 100                       | 100                                 |
| **Печенье**  |  53           | 53                        | 53                                  |
| **Злаки**    |  0            | 0                         | 0                                   |
| **Овощи**    |  143          | 143                       | 143                                 |
| **Масло**    |  20           | 0                         | 20                                  |
| **Сыр**      |  154          | 113                       | 0                                   |
| **Молоко**   |  378          | 378                       | 378                                 |
| **Орехи**    |  15           | 0                         | 15                                  |
| **Рыба**     |  0            | 0                         | 0                                   |
| **Мясо**     |  317          | 317                       | 44                                  |


Все данные свел в итоговую таблицу для облегчения итогового анализа.

1. Можем заметить, что во втором новом рационе вернулись **яйца**, **масло** и **орехи** в том же количестве, что и раннее.
2. Сыр вовсе вышел из рациона
3. Потребления мяса снизилось на 273 грамма
4. Рыба в рацион так и не добавилась


# 3. Провести анализ чувствительности рациона к стоимости и порогу на выброс парниковых газов.

Для начала проведем анализ чувствительность рациона на выбросы. Их будем увеличивать незначительно. На 0.1кг

In [19]:
db = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
prew_solution = linprog(c, A_ub, b_ub, method='simplex')['fun']
a= 1
while(True):
    db[23] += 0.1
    b_ub[23]+=db[23]
    solution = linprog(c, A_ub, b_ub, method='simplex')
    if (solution['status']!=0):
        print('Couldn''t find a solution')
        break;
    new_solution = solution['fun']
    delta = new_solution-prew_solution
    print(a, " Итерация: ", new_solution, "delta: ", delta)
    if (abs(delta)) < 0.1e-10:
        print("Завершение")
        break;
    prew_solution = new_solution
    a+=1

1  Итерация:  357.24195804196074 delta:  -13.986013986007208
2  Итерация:  329.2699300699348 delta:  -27.972027972025955
3  Итерация:  287.31188811188633 delta:  -41.958041958048454
4  Итерация:  232.69188164204755 delta:  -54.620006469838785
5  Итерация:  189.751566601501 delta:  -42.94031504054655
6  Итерация:  180.18646568979813 delta:  -9.56510091170287
7  Итерация:  174.78777777777628 delta:  -5.398687912021842
8  Итерация:  174.78777777778362 delta:  7.332801033044234e-12
Завершение


Аналогично проведем анализ на стоимость. Здесь увеличение на 1

In [20]:
db = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
prew_solution = linprog(c, A_ub, b_ub, method='simplex')['fun']
a= 1
while(True):
    db[22] += 1
    b_ub[22]+=db[22]
    solution = linprog(c, A_ub, b_ub, method='simplex')
    if (solution['status']!=0):
        print('Couldn''t find a solution')
        break;
    new_solution = solution['fun']
    delta = new_solution-prew_solution
    print(a, " Итерация: ", new_solution, "delta: ", delta)
    if (abs(delta)) < 0.1:
        print("Завершение")
        break;
    prew_solution = new_solution
    a+=1

1  Итерация:  173.3592063492124 delta:  -1.4285714285712174
2  Итерация:  170.5020634920695 delta:  -2.8571428571428896
3  Итерация:  166.21634920635535 delta:  -4.285714285714164
4  Итерация:  160.50206349206954 delta:  -5.714285714285808
5  Итерация:  153.35920634921254 delta:  -7.142857142856997
6  Итерация:  144.78777777779342 delta:  -8.57142857141912
7  Итерация:  135.5702173774546 delta:  -9.217560400338812
8  Итерация:  125.37068910562766 delta:  -10.19952827182695
9  Итерация:  113.89621979984071 delta:  -11.474469305786954
10  Итерация:  101.14680946006447 delta:  -12.749410339776233
11  Итерация:  87.12245808631543 delta:  -14.024351373749042
12  Итерация:  71.82316567858936 delta:  -15.299292407726071
13  Итерация:  55.24893223687261 delta:  -16.57423344171675
14  Итерация:  37.39975776120346 delta:  -17.84917447566915
15  Итерация:  21.91850104821806 delta:  -15.481256712985399
16  Итерация:  9.775999999996559 delta:  -12.142501048221503
17  Итерация:  0.4033613445388369 d